### PR TITLE
major: full dynamic pages and plain next.js

### LINKED ISSUE #5404: Redirection of Content through SSR/Middlewares
With the addition of https://github.com/nodejs/nodejs.org/pull/5396, we can finally have custom middleware that can be fired to handle different aspects of our website.

With the addition of https://github.com/nodejs/nodejs.org/pull/5397/, we can have a single 404 (Not Found) page instead of one per `pages/{languageCode}/404.md`. This is useful for static generation since it uses Localization data now backed directly within the Application, instead of the old way where Localization data is injected into the App's individual `pageProps`. This creates some advantages, and it's a foundational step (together with the new middleware engine) that allows us to cover the functionality of this issue.

### What is this issue about?

The Node.js Website is entirely static. We now support SSR and have some cool Server-Side features on the pipeline (such as search), but fundamentally, the pages are static.

This means that each page (which resides in each language folder) is a single unit. This is done since we want to support (for the long term) the website to fully operate on a static basis without requiring any server or server-side routing. Next.js's built-in i18n feature requires a single folder within `pages` (e.g. /pages/blog/index.md, /pages/index.md, /pages/about/something.md) instead of subfolder localised paths (e.g. /pages/en/blog/index.md, etc.)

The main issue is that Next.js built-in i18n routing could bring Translation Contexts and do the Routing for us. With https://github.com/nodejs/nodejs.org/pull/5397/ and the current implementation, we pretty much have an in-house Localisation system. The main benefit of the current PR is that the Translation system is not static and bounded during build-time, which already brings benefits, such as for the 404 page.

### The big blocker

In other words, our next big blocker is how we handle loading pages that do not exist in a specific language.

With https://github.com/nodejs/nodejs.org/pull/5397/, loading Translation Context and Localized Messages is not an issue anymore, as the messages are eval'd during both build-time and runtime.

Yet, we have one big problem **How can we tell Next.js that once it encounters a 404 to attempt to load a page with a similar path (e.g.,/de/blog/something-something we want to extract /blog/something-something) and load the source page (English version) of it if exists?

- These should not be done/generated during build-time.
  - It's OK not to support these with static builds, but it's doable. We need to use `getStaticPaths` and `getStaticProps`

### The technical hurdle

Next.js supports dynamic pages (even with Static Generation-only, aka "Static Dynamic Pages") (https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes, https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-paths)

This could allow us to do something like `pages/[language]/[...path].tsx` (or `.mdx` if we want to make it dynamic-static).

Until then, all good; we're able to dynamically handle pages, even tho if we use the static approach, we would need to make beforehand a complete list of all pages that each language does not have (a sort of a diff), which by itself seems like an over-engineered solution.

**Then we have an even harder blocker:** How do we load the English version of the page within an already computed page? (As far as I know, there's no way to do a dynamic import that loads pages with a dynamically constructed path from a `prop`).

### Open to ideas

This is not necessarily a blocker for launching the nodejs.org website on Vercel (as the same functionality is not on the https://nodejs.org website right now, as it got removed once we moved from Handlebars/Metalsmith). Still, it is a critically important feature in the long-term run.

It'd be ideal if Next.js Middlewares had a way to know if the requested route exists or not and to be able to redirect to another page with an injected context (?) (cc @leerob @Ethan-Arrowood @shuding)

Or any other way that we could tell Next.js to load an alternative page... Well, @nodejs/website I'm open for suggestions :)

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR implements the [Discussion started here](https://github.com/nodejs/nodejs.org/discussions/5442) on leaving Nextra in favour of a vanilla Next.js experience.

Truth be told, as mentioned in the discussion, we finally matured so that we don't need Nextra anymore.

This PR implements an engine that allows Incremental Builds (SSR) and full Static Builds of non-localised pages with localised context(s).

Leveraging Next.js Dynamic Routes feature (Static Paths + On-demand Static Page Generation).

This also uses [MDX Remote](https://github.com/hashicorp/next-mdx-remote) for the compilation of MDXm for Next.js on build-time and on-demand.

### How it works

- We first detect if the built environment is "static export" or regular builds (Standalone, Vercel)
- We determine if the pages should be built statically or not based on that (since with standalone/vercel we can support complete ISR for all existing pages
  - If on an ISR environment, we also support generating blog pages on-demand
  - Otherwise, we only generate all pages, excluding blog posts, as the static builds would be gigantic and too long.
  - On the ISR environment, we skip building the non-localized pages at build time to speed up the build.
- Then, when the page is requested, we load their Markdown data and then parse the Markdown source with `MDXRemote`
  - We add extra context (such as blog data) if needed
- The page is rendered by invoking the `NextraTheme` with necessary props

## Related Issues

- Closes #5404

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [x] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Comment by AugustinMauroy** [comment]
Hey! That's great.
Question do we keep the contents in the page folder? Wouldn't it be smarter to make a content folder? So in the `pages` folder we'd only have `.tsx` and the `md|mdx` would be in the `content` directory. The `pages` folder would be the skeleton with the layouts.

**Comment by ovflowd** [comment]
> Question do we keep the contents in the page folder? Wouldn't it be smarter to make a content folder? So in the pages folder we'd only have .tsx and the md|mdx would be in the content directory. The pages folder would be the skeleton with the layouts.

No, it is not smarter to have a `content` folder. As I mentioned before and it was documented before, for the time being, we want to keep the structure of the content/pages as much as simple and easy to understand for existing long-term contributors and newcomers.

A specific content folder brings no benefits and only adds complexity. A content folder also doesn't solve the original issue, so not even sure why you're mentioning it here.

> The pages folder would be the skeleton with the layouts.

I know how this works, we did it on Gatsby a long time ago. But Next.js works differently, and even tho we could load the content dynamically only from a content folder, it is kinda anti-intuitive.

The current approach exists to support non-localized pages to have their source content to be shown and still be localised.

**This works because:**

- We know all the paths at build time by analysing the `pages/en/**` folder, with the content folder we have no idea what is the SLUG that each page should have. (Only if we adopted the same folder structure)
- We don't want to have dynamic page creation for static pages. This is an anti-pattern. The biggest purpose for dynamic pages is at the title itself "dynamic pages", not "static pages". The non-localized pages fit this criterion because they are actually dynamic because they don't exist and might change at each build (as more content gets translated).
- We want to avoid having a separation of content <> page skeleton. A page should have its content and components together. This would avoid the mess we had on Gatsby.

Also, gentle reminder that you already asked me this a few times, and I find it kinda annoying that you keep insisting on it (if that was not your intention, then I'm sorry 😅)

**Review by AugustinMauroy — APPROVED** [review]
LGTM !
Thanks Claudio for your answer.

**Comment by ovflowd** [comment]
### Updates on the PR

- I've updated this PR replacing some of the existing logic for the `next-data` middlewares. Also note that once #5365 gets merged `next-data` will pretty much only be generating runtime blog data for the blog pages and the RSS feeds rebuild. 
- The existing `next-data` folder was moved to the root `next-data` and the files were updated with better docs, removal of unused code, improvement of existing code
- The PR now introduces dynamic routes for `blog/year-XXXX` routes, instead of generating Markdown on the filesystem
- Some other minor improvements were done to performance
- Files were relinted (hence some changes purely on linting) and reformatted

**Comment by ovflowd** [comment]
cc @nodejs/website having reviews would be nice here :)

File: next-data/generateWebsiteFeeds.mjs [hunk 1]
```diff
@@ -1,37 +1,28 @@
...
 
-export const generateBlogYearPages = cachedBlogData =>
-  cachedBlogData('', true)
-    .then(({ blogData }) => blogData.posts.map(p => p.date.getFullYear()))
-    .then(data => [...new Set(data)])
-    .then(data => data.forEach(createBlogYearFile));
-
-export const generateWebsiteFeeds = cachedBlogData =>
+/**
+ * This method generates RSS website Feeds based on the current Website Configuration
```
Comment (bmuenzenmeyer) [inline next-data/generateWebsiteFeeds.mjs]: nit: the casing of `Feeds` and `Website Configuration` threw me off and made me wonder if they were variables for a moment

```suggestion
 * This method generates RSS website feeds based on the current website configuration
```

File: next-data/getBlogData.mjs [hunk 2]
```diff
@@ -33,23 +31,39 @@ const getPost = category => post =>
...
+    .getDirectories(__dirname, blogPath)
+    .then(c => c.map(s => [s, readdir(join(blogPath, s))]));
 
   const categoriesPosts = blogCategories.then(c =>
     c.map(([s, f]) => [s, getPosts(s, f)])
   );
 
-  return (route = '/', getAllAvailablePosts = false) => {
+  /**
+   * This method generates the Blog Post data based on route
```
Comment (bmuenzenmeyer) [inline next-data/getBlogData.mjs]: nit: again, casing of `Blog` and `Post`

Comment (bmuenzenmeyer) [reply next-data/getBlogData.mjs]:
```suggestion
   * This method generates the blog post data based on route
```

File: next-data/helpers.mjs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import { existsSync } from 'node:fs';
+import { fileURLToPath } from 'node:url';
+import { glob } from 'glob';
+
+export const getMatchingRoutes = (route = '', matches = []) =>
+  matches.some(match => route === match);
+
+/**
+ * This methos is responsible for reading all immediate subdirectories of a directory
```
Comment (bmuenzenmeyer) [inline next-data/helpers.mjs]: typo: `methos`

```suggestion
 * This method is responsible for reading all immediate subdirectories of a directory
```

File: next-data/helpers.mjs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * @returns {Promise<string[]>} a promise containing an array of directories
+ */
+export const getDirectories = async (root, cwd) => {
+  return glob('*', { root, cwd, withFileTypes: true })
+    .then(d => d.filter(e => e.isDirectory()))
+    .then(d => d.map(e => e.name));
+};
+
+/**
+ * This get ths relative path from `import.meta.url`
```
Comment (bmuenzenmeyer) [inline next-data/helpers.mjs]: typo: `ths`
```suggestion
 * This gets the relative path from `import.meta.url`
```

File: next-data/helpers.mjs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * @returns {Promise<string[]>} a promise containing an array of paths
+ */
+export const getMarkdownFiles = async (root, cwd, ignore = []) => {
+  return glob('**/*.{md,mdx}', { root, cwd, ignore })
+    .then(files => files.map(file => file.replace(/(\/index)?\.mdx?$/, '')))
+    .then(files => files.filter(file => file.length));
+};
+
+/**
+ * This method is responsible by checking a combination
```
Comment (bmuenzenmeyer) [inline next-data/helpers.mjs]: typo: `by`

```suggestion
 * This method is responsible for checking a combination
```

File: next-data/helpers.mjs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * If no matching extension is found an empty string is returned
+ * which signals that the find was not found
+ *
+ * @param {string} filename the filename without extension/suffixes
+ * @param {string[]} extensions an array of suffixes to be tested
+ * @returns {string} the filename with the first matching extension
+ */
+export const checkFileExists = (filename, extensions) => {
+  const extension = extensions.filter(e => existsSync(`${filename}${e}`))[0];
```
Comment (bmuenzenmeyer) [inline next-data/helpers.mjs]: I wonder if this should be a [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) instead

File: next.dynamic.mjs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+const __dirname = nextData.helpers.getRelativePath(import.meta.url);
+
+// generates the current blog data
+const cachedBlogData = nextData.getBlogData();
+
+// this retrieves all pages that are under the english locale directory
+// besides blog posts since they should not be localised at all
+const sourcePages = await nextData.helpers.getMarkdownFiles(
+  __dirname,
+  `pages/${nextLocale.defaultLocale.code}`,
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: I started marking up this file with `path.sep` comments and then re-read the docs and it seems they are now valid Windows separators during reads, as long as you know that the path method writes will follow `sep` logic.

File: next.dynamic.mjs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+ * As a security measure we also check against `validFallbackFolders` to ensure that at the last scenario
+ * the base path comes from a valid base folder. Next.js will already protect against common attack vectors
+ * such as `/../../` on the URL pathname and other methodologies
+ *
+ * @param {string[]} paths the pathname string as an array (split by /)
+ * @returns {{ source: string, filename: string, pathname: string }} the source and filename
+ * @throws {Error} if the file does not exist, which should never happen
+ */
+export const getMarkdownFileSource = paths => {
+  const pathname = paths.join('/');
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: contrary to my other comment, this ___might___ not be Windows-safe.

File: next.dynamic.mjs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+  return { source: '', filename: '', pathname };
+};
+
+/**
+ * This Method gathers the Markdown Source and the source filename
+ * and processes the data (parses the markdown) and generate props
+ * for the application to consume (`getStaticProps`)
+ *
+ * @param {string} source the Mardkown source file
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: typo `Mardkown`

```suggestion
 * @param {string} source the Markdown source file
```

File: pages/[locale]/[...pathname].tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  StaticParams
+> = async ({ params = { pathname: [], locale: 'en' } }) => {
+  // We retrieve the source of the Markdown file by doing an educated guess
+  // of what possible files could be the source of the page, since the extension
+  // context is lost from `getStaticProps` as a limitation of Next.js itself
+  const { source, filename, pathname } = nextDynamic.getMarkdownFileSource(
+    params.pathname
+  );
+
+  // This parsers the actual Markdown content and returns a full set of props
```
Comment (bmuenzenmeyer) [inline pages/[locale]/[...pathname].tsx]: typo: `parsers`

```suggestion
  // This parses the actual Markdown content and returns a full set of props
```

File: pages/[locale]/blog/[year].tsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+> = async ({ params = { year: '' } }) => {
+  // We replace the `year-` to get the actual year number that we need
+  const yearNumber = params.year.replace('year-', '');
+
+  // We generate the source of the Markdown file based on how a "year" page should look like
+  const source = availableYears.includes(yearNumber)
+    ? `---\nlayout: blog-index.hbs\ntitle: News from ${yearNumber}\npaginate: blog\n---\n`
+    : '';
+
+  // This parsers the actual Markdown content and returns a full set of props
```
Comment (bmuenzenmeyer) [inline pages/[locale]/blog/[year].tsx]: typo: `parsers`

```suggestion
  // This parses the actual Markdown content and returns a full set of props
```

File: theme.dynamic.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      pageMap: [],
+      title: frontMatter.title || 'Node.js',
+      filePath: '',
+      route: '',
+    }),
+    [frontMatter]
+  );
+
+  return (
+    <NextraTheme pageOpts={pageOpts} themeConfig={null} pageProps={props}>
```
Comment (bmuenzenmeyer) [inline theme.dynamic.tsx]: I don't get why the `themeConfig` is null here, but I am still learning about custom themes in nextra and comingling next with it. Perhaps it warrants a comment? Disregard if you wish.

**Review by bmuenzenmeyer — COMMENTED** [review]
An initial manual pass - hoping to run the code this weekend on my Windows machine

File: next-data/helpers.mjs (hunk 6 above)
Comment (ovflowd) [reply next-data/helpers.mjs]: That makes sense!

File: next.dynamic.mjs (hunk 8 above)
Comment (ovflowd) [reply next.dynamic.mjs]: Well, we can definitely test this on Windows, whenever you got some time. One good way of knowing if this worked is by running `npx turbo deploy` on your Windows machine.

Comment (bmuenzenmeyer) [reply next.dynamic.mjs]: Initial results seem to be a failure somewhere else. I don't want to suggest the PR until I can learn more - perhaps it's my environment.

image

**Comment by ovflowd** [comment]
Going to rebase and update based on code-review soon.

File: next.dynamic.mjs (hunk 8 above)
Comment (ovflowd) [reply next.dynamic.mjs]: I saw the `_isAppDir` error happening also on Windows Builds on GitHub Actions but it solved by itself.

Maybe if you clear the Next.js cache or something?

Comment (ovflowd) [reply next.dynamic.mjs]: Waiting for support from Vercel https://github.com/vercel/next.js/issues/51225

Comment (ovflowd) [reply next.dynamic.mjs]: Important to mention that the GitHub Actions is passing now (https://github.com/nodejs/nodejs.org/actions/runs/5261348136/jobs/9509358742?pr=5426) for this branch so who knows...

Comment (bmuenzenmeyer) [reply next.dynamic.mjs]: thanks for checking into it. I am not too concerned if CI is working well

File: theme.dynamic.tsx (hunk 12 above)
Comment (ovflowd) [reply theme.dynamic.tsx]: Usually `themeConfig` is used for built-in Nextra Themes, doesn't seem to be something that needs to be used on Custom Themes.

I bet that on the actual usage when directly imported by `theme.tsx` it's also null. But I can leave a comment, yes.

**Comment by ovflowd** [comment]
FYI I've opened an Issue on Vercel (#00142137) as it seems for some reason 5XX errors are happening on the Vercel for dynamic pages (Serverless Functions)

Errors such as:

```
Error: > Couldn't find a `pages` directory. Please create one under the project root
    at findPagesDir (/var/task/node_modules/next/dist/lib/find-pages-dir.js:104:19)
    at findPagesDirectory (/var/task/node_modules/nextra/dist/index.js:155:15)
    at Object.<anonymous> (/var/task/node_modules/nextra/dist/index.js:159:53)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at 40850 (/var/task/.next/server/pages/[locale]/[...pathname].js:341:18) {
  page: '/ca/get-involved/contribute'
}
RequestId: 428a84fe-b44d-42bc-802f-219973206b34 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

This tells us that the `pages` directory doesn't exist when it clearly does...

Note that this works normally/fine on local production builds.

**Comment by ovflowd** [comment]
**Note.:** This probably cannot get merged until we fix the issue on Vercel.

**Comment by ovflowd** [comment]
Hey, @nodejs/website I've updated this PR and now it contains the full solution of:

- Using plain Next.js without Nextra
- Dynamic Pages
- Refactoring of `next-data`

Everything is working as expected now!

**Review by araujogui — APPROVED** [review]
LGTM

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -60,60 +58,60 @@ for getting things done and landing your contribution.
...
 2. Clone your fork using SSH, GitHub CLI, or HTTPS.
 
-    ```bash
-    git clone git@github.com:<YOUR_GITHUB_USERNAME>/nodejs.org.git # SSH
-    git clone https://github.com/<GITHUB_ID>/nodejs.org.git # HTTPS
-    gh repo clone <GITHUB_ID>/nodejs.org # GitHub CLI
-    ```
+   ```bash
+   git clone git@github.com:<YOUR_GITHUB_USERNAME>/nodejs.org.git # SSH
+   git clone https://github.com/<GITHUB_ID>/nodejs.org.git # HTTPS
```
Comment (bmuenzenmeyer) [inline CONTRIBUTING.md]: making this consistent with a suggestion from https://github.com/nodejs/nodejs.org/pull/5441

```suggestion
   git clone https://github.com/<YOUR_GITHUB_USERNAME>/nodejs.org.git # HTTPS
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -60,60 +58,60 @@ for getting things done and landing your contribution.
...
+   gh repo clone <GITHUB_ID>/nodejs.org # GitHub CLI
```
Comment (bmuenzenmeyer) [inline CONTRIBUTING.md]: making this consistent with a suggestion from https://github.com/nodejs/nodejs.org/pull/5441


```suggestion
   gh repo clone <YOUR_GITHUB_USERNAME>/nodejs.org # GitHub CLI
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -186,7 +184,7 @@ We recommend a read on our [Collaborator Guide](COLLABORATOR_GUIDE.md#accepting-
     - The person that is fast-tracking the PR (adding the label) must also comment on the PR that they're requesting the PR to be fast-tracked
     - The comment must mention `@nodejs/website` and must have at least one 👍 (or any other sort of approval reaction) if the person fast-tracking the PR is the author of the PR.
     - Fast-tracking is only allowed for small bug fixes, small feature changes, localisation changes, or other sorts of non-critical/highly-impacting changes that are not covered by the previous rule that allows PRs to be merged immediately.
-    - Fast-tracking cannot be used for updates on the `COLLABORATOR_GUIDE.md`,  CONTRIBUTING.md` guide, `CODEOWNERS`, GitHub Actions or any security-impacting file or document that changes the governing policies of this repository.
+    - Fast-tracking cannot be used for updates on the `COLLABORATOR_GUIDE.md`, CONTRIBUTING.md`guide,`CODEOWNERS`, GitHub Actions or any security-impacting file or document that changes the governing policies of this repository.
```
Comment (bmuenzenmeyer) [inline CONTRIBUTING.md]: the backticks on this file are wrong, leading to incorrect rendering of the inline code block

image

```suggestion
    - Fast-tracking cannot be used for updates on the `COLLABORATOR_GUIDE.md`, `CONTRIBUTING.md` guide, `CODEOWNERS`, GitHub Actions, or any security-impacting file or document that changes the governing policies of this repository.
```

File: components/Home/HomeDownloadButton.tsx [hunk 16]
```diff
@@ -14,15 +14,16 @@ const HomeDownloadButton: FC<NodeRelease> = ({
...
 
   const { os, bitness } = useDetectOS();
 
   const nodeDownloadLink = downloadUrlByOS(versionWithPrefix, os, bitness);
   const nodeApiLink = `https://nodejs.org/dist/latest-v${major}.x/docs/api/`;
   const nodeAllDownloadsLink = `/download${isLts ? '/' : '/current'}`;
   const nodeDownloadTitle =
-    `${labels.download} ${version}` + ` ${labels[isLts ? 'lts' : 'current']}`;
+    `${labels?.download} ${version}` +
+    ` ${labels?.[isLts ? 'lts' : 'current']}`;
```
Comment (bmuenzenmeyer) [inline components/Home/HomeDownloadButton.tsx:26]: nit: string interpolation and concatenation here. I could see it be purposeful for readability, but suggesting a change anyway to see what we prefer. I didn't check, perhaps it's a line length thing with prettier too. feel free to disregard if desired.

```suggestion
    `${labels?.download} ${version} ${labels?.[isLts ? 'lts' : 'current']}`;
```

File: components/Home/HomeDownloadButton.tsx [hunk 17]
```diff
@@ -32,23 +33,23 @@ const HomeDownloadButton: FC<NodeRelease> = ({
         title={nodeDownloadTitle}
         data-version={versionWithPrefix}
       >
-        {version} {labels[isLts ? 'lts' : 'current']}
-        <small>{labels[`tagline-${isLts ? 'lts' : 'current'}`]}</small>
+        {version} {labels?.[isLts ? 'lts' : 'current']}
```
Comment (bmuenzenmeyer) [inline components/Home/HomeDownloadButton.tsx:36]: so many optional chains in this file make me wonder if it should be short circuited once at the top level?

File: hooks/useBlogData.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+  const getPagination = useCallback(
+    (currentYear: number) => ({
+      next: pagination.includes(currentYear + 1) ? currentYear + 1 : undefined,
+      prev: pagination.includes(currentYear - 1) ? currentYear - 1 : undefined,
+    }),
+    [pagination]
+  );
+
+  const currentCategory = useMemo(
+    () => asPath.split('/')[3] || new Date().getFullYear().toString(),
```
Comment (bmuenzenmeyer) [inline hooks/useBlogData.ts]: nit: magic index here worthy of a comment

File: next-data/generateBlogPostsData.mjs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    ['**/index.md', '**/pagination.md']
+  );
+
+  // we gather all the information of all the blog posts by reading each individual file
+  // and then parsing the frontmatter and source content and returning a minified object
+  const postsPromise = filenames
+    .map(name => ({ name, file: readFile(join(blogPath, name)) }))
+    .map(({ name, file }) => file.then(source => getFrontMatter(name, source)));
+
+  // we await for all the work to be conclued and return a nice blog posts object
```
Comment (bmuenzenmeyer) [inline next-data/generateBlogPostsData.mjs]: typo: `conclued`

```suggestion
  // we await for all the work to be concluded and return a nice blog posts object
```

File: next.constants.mjs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * This is used for static/without a Node.js server hosting, such as on our
+ * legacy Website Build Environment on Node.js's DigitalOcean Droplet.
+ */
+export const ENABLE_STATIC_EXPORT =
+  process.env.NEXT_STATIC_EXPORT === 'true' ||
+  process.env.NEXT_STATIC_EXPORT === true;
+
+/**
+ * Supports a manuall override of the base path of the Website
```
Comment (bmuenzenmeyer) [inline next.constants.mjs]: typo: `manuall`

```suggestion
 * Supports a manual override of the base path of the Website
```

File: next.dynamic.mjs [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+
+// allows us to run a glob to get markdown files based on a language folder
+const getPathsByLanguage = async (
+  locale = nextConstants.DEFAULT_LOCALE_CODE,
+  ignored = []
+) =>
+  nextData.helpers.getMarkdownFiles(process.cwd(), `pages/${locale}`, ignored);
+
+/**
+ * This method is responsible for generating a Collection of all availalbe paths that
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: typo: `available`

```suggestion
 * This method is responsible for generating a Collection of all available paths that
```

File: next.dynamic.mjs [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+ * are served by the Website dynamically based on the Markdown pages on `pages/` folder.
+ *
+ * Each Collection is associated to its Locale Code and containins a subset of Dictionaries
```
Comment (bmuenzenmeyer) [inline next.dynamic.mjs]: typo: `containins`

```suggestion
 * Each Collection is associated to its Locale Code and contains a subset of Dictionaries
```

File: next.locales.mjs [hunk 23]
```diff
@@ -46,8 +45,8 @@ const getCurrentTranslations = locale => ({
 });
 
 export {
-  availableLocales,
   defaultLocale,
+  availableLocales,
```
Comment (bmuenzenmeyer) [inline next.locales.mjs]: nit: unneeded diff, and now not alphabetical

File: pages/[...path].tsx [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+  const { source, filename } = nextDynamic.getMarkdownFile(
+    locale,
+    rewriteRule ? rewriteRule(pathname) : pathname
+  );
+
+  // This parses the actual Markdown content and returns a full set of props
+  // to be passed to the base page (`DynamicPage`) which will render the Markdown
+  const staticProps = await nextDynamic.getStaticProps(source, filename);
+
+  // This checks if either we already the route does not exist or if we should ignore
```
Comment (bmuenzenmeyer) [inline pages/[...path].tsx]: nit grammar:

```suggestion
  // This checks if either we already determined the route does not exist or if we should ignore
```

**Review by bmuenzenmeyer — COMMENTED** [review]
I want to run the code locally again but thought I'd submit this initial review

File: components/Home/HomeDownloadButton.tsx (hunk 17 above)
Comment (ovflowd) [reply components/Home/HomeDownloadButton.tsx:36]: The optional chains here are just to make TypeScript happy. Remember this file is going to be removed soon together with all the legacy templates and pages.

For this page this frontmatter property (labels) is always present

File: components/Home/HomeDownloadButton.tsx (hunk 16 above)
Comment (ovflowd) [reply components/Home/HomeDownloadButton.tsx:26]: I would avoid making any further improvements to these files as they will be removed soon. They're legacy files.

**Review by aymen94 — APPROVED** [review]

**Review by bmuenzenmeyer — APPROVED** [review]

**Comment by ovflowd** [comment]
@aymen94 I think it was premature to get such a massive PR merged without more reviews from other folks.

But I guess it's fine 😅

**Comment by aymen94** [comment]
> @aymen94 I think it was premature to get such a massive PR merged without more reviews from other folks.
> 
> But I guess it's fine 😅

@ovflowd I understand your point regarding the number of reviews for merging a completely new development. While four reviews may seem like enough, I appreciate your suggestion to pay attention. 😁

**Pull request merged**